### PR TITLE
Dataengine

### PR DESCRIPTION
**Pull request by Charithaakula**
Adding Data Engine

### CODE REVIEW

File: dataengine/generate_qa.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+            p.join()
+        processed_data.extend(results)
+
+        with open(processed_data_file_path, 'w') as f:
+            json.dump(list(processed_data), f, indent=4)
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Process some topics.')
+    parser.add_argument('--fields', nargs='+', default=[
```
[the remark this answers is not in the document]
Comment (Charithaakula) [reply dataengine/generate_qa.py:227]: As of now, yes. But the script can read from topics folder potentially.

File: dataengine/generate_vqa.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    # Optionally, rename the files in the directory (uncomment to use)
+    for original_filename, new_filename in renamed_files.items():
+        os.rename(os.path.join(image_dir, original_filename),
+                  os.path.join(image_dir, new_filename))
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description='Process and rename images and update JSON data accordingly.')
+    parser.add_argument('--topicname', type=str,
```
[the remark this answers is not in the document]
Comment (Charithaakula) [reply dataengine/generate_vqa.py:55]: Yeah, it's a fairly light post processing script, thus didn't parallelize it. I use to run this script in a single bash with multiple topics but potentially can take all topics at once.

File: dataengine/README.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+The workflow consists of a series of Python scripts that should be executed in the following order in the appropriate environment (see requirements.txt):
+
+```bash
+#!/bin/bash
+
+# set your environment / keys
+OPENAI_API_KEY="your_openai_key"
+GOOGLE_API_KEY="your_google_api_key"
+GOOGLE_SE_ID="your_google_search_engine_id"
+USER_AGENT="your_user_agent"
```
[the remark this answers is not in the document]
Comment (Charithaakula) [reply dataengine/README.md:21]: I used this - 'Image downloader/1.0 (sca321@nyu.edu)' but the above would work too. Maybe we can add the reference in the comment in the USER_AGENT line?

[the remark this answers is not in the document]
Comment (ellisbrown) [reply dataengine/README.md:21]: @Charithaakula couple qs:

1. this is different than the wikipedia user agent (line below) right?
2. is this downloading images from google or from wikipedia?
   1. if from wikipedia: we could just use the same `WIKIPEDIA_USER_AGENT` right
   2. if from google: any reason you need to add your email?
      - if not, we might consider simplifying it and just using a standard one like I posted that just mimics a browser?
      - if there is some reason to include custom client name / email, maybe just add a note about what is required?

File: dataengine/generate_qa.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ellisbrown) [reply dataengine/generate_qa.py:227]: I think we have two options:

1. quickly modify the script to read topics from topic folder or from the input fields txt
2. manually pass a topic in the example block on the readme

I am fine with either. just is a bit unclear without this

File: dataengine/generate_vqa.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ellisbrown) [reply dataengine/generate_vqa.py:55]: same as above. no need to parallelize / speed up. just want the demonstration script to be "feature complete" & for people to understand what all is required to replicate

**Review by tsb0601 — APPROVED** [review]
I think the code looks very good to me. The readme should be clear to follow

File: dataengine/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Charithaakula) [reply dataengine/README.md:21]: It's generally recommended that when scraping large amounts of data, to retain accountability, we need to use our personal email, hence I went for my email, maybe it's best if we even give the same as example?

**Pull request merged**